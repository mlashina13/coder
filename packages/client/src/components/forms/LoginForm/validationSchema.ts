import * as yup from 'yup';

/**
 * Схема валидации
 */
export const validationSchema = yup.object({
  login: yup.string().required('Введите логин'),
  password: yup.string().required('Введите пароль'),
});
