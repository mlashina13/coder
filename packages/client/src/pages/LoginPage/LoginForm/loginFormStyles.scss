@import '../../../styles/color';
@import '../../../styles/common';

.login-form {
  display: flex;
  flex-direction: column;
  gap: $spacing-ms;

  &__input {
    width: 25rem;
  }
}
