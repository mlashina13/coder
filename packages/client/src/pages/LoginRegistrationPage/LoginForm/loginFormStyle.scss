@import '../../../styles/color';
@import '../../../styles/common';

.login-form {
  display: flex;
  flex-direction: column;
  gap: $spacing-ms;

  &__input {
    width: 25rem;
  }

  &__link-btn {
    &.MuiButton-text {
      color: #0E4980;
      text-align: center;
      font-style: normal;
    }
  }
}
