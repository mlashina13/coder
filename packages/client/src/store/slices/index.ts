export { errorReducer } from './errorSlice';
export { userReducer } from './userSlice';
export { locationReducer } from './locationSlice';
export { leadboardReducer } from './leaderboardSlice';
export { oauthReducer } from './oauthSlice';
