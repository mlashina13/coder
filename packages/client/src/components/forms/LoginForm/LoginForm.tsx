import { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useFormik } from 'formik';
import { Button, Input } from '../../common';
import { LoginData } from '../../../types/common';
import { validationSchema } from './validationSchema';
import { useAppDispatch, useAppSelector } from '../../../hooks';
import { findLocation, login } from '../../../services';
import { ROUTER_URLS } from '../../../constants';
import './loginFormStyles.scss';

/**
 * Форма авторизации
 */
export const LoginForm = () => {
  const { loading, currentUser } = useAppSelector((state) => state.user);
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (currentUser) {
      navigate(ROUTER_URLS.Main);
    }
  }, [currentUser]);

  const formik = useFormik<LoginData>({
    initialValues: {
      login: '',
      password: '',
    },
    validationSchema,
    onSubmit: (values) => {
      dispatch(login(values));
      dispatch(findLocation());
    },
    validateOnBlur: false,
  });

  return (
    <form onSubmit={formik.handleSubmit} className='login-form'>
      <Input
        name='login'
        variant='filled'
        value={formik.values.login}
        onChange={formik.handleChange}
        error={formik.touched.login && Boolean(formik.errors.login)}
        helperText={formik.touched.login && formik.errors.login}
        className='login-form__input requiredField'
        label='Логин'
        autoComplete='off'
        size='small'
      />
      <Input
        name='password'
        variant='filled'
        value={formik.values.password}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.password && Boolean(formik.errors.password)}
        helperText={formik.touched.password && formik.errors.password}
        type='password'
        className='login-form__input requiredField'
        label='Пароль'
        autoComplete='off'
        size='small'
      />
      <Button label='Войти' type='submit' disabled={loading} />
    </form>
  );
};
