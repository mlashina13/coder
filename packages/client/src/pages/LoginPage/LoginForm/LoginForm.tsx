import { useFormik } from 'formik';
import * as yup from 'yup';

import { Button, Input } from '../../../components';
import { Auth } from '../../../services';
import { useUserStore } from '../../../stores';

import './loginFormStyles.scss';

/** схема валидации */
const validationSchema = yup.object({
  login: yup
    .string()
    /** проверка через регулярное выражение логина */
    .matches(/^[a-zA-Z0-9-_]{3,20}$/g, {
      message:
        'Логин должен быть от 3 до 20 символов, написан латиницей, допускаются цифры, дефис и нижнее подчёркивание',
    })
    /** указываем на обязательность логина */
    .required('Введите логин'),
  password: yup
    .string()
    /** проверка через регулярное выражение пароля */
    .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)/g, {
      message:
        'Пароль должен быть от 8 до 40 символов, обязательно хотя бы одна заглавная буква и одна цифра',
    })
    /** указываем на обязательность пароля */
    .required('Введите пароль'),
});

/** форма авторизации */
export const LoginForm = () => {
  const { setUserData } = useUserStore();
  const formik = useFormik({
    initialValues: {
      login: '',
      password: '',
    },
    validationSchema,
    onSubmit: (values) => {
      Auth.signIn(values).then(() => {
        Auth.getAuthUser().then((data) => setUserData?.(data));
      });
    },
    validateOnBlur: false,
  });
  return (
    <form onSubmit={formik.handleSubmit} className='login-form'>
      <Input
        name='login'
        variant='filled'
        value={formik.values.login}
        onChange={formik.handleChange}
        error={formik.touched.login && Boolean(formik.errors.login)}
        helperText={formik.touched.login && formik.errors.login}
        className='login-form__input requiredField'
        label='Логин'
        autoComplete='off'
        size='small'
      />
      <Input
        name='password'
        variant='filled'
        value={formik.values.password}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.password && Boolean(formik.errors.password)}
        helperText={formik.touched.password && formik.errors.password}
        type='password'
        className='login-form__input requiredField'
        label='Пароль'
        autoComplete='off'
        size='small'
      />
      <Button label='Войти' type='submit' />
    </form>
  );
};
