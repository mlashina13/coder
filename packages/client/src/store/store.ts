import { configureStore } from '@reduxjs/toolkit';
import { errorReducer, userReducer, locationReducer } from './slices';

declare global {
  interface Window {
    // В d.ts нам неважно, что это за тип,
    // так как он сразу попадает в redux store на клиенте
    __PRELOADED_STATE__?: object;
  }
}

// console.log(serverState);

/**
 * Хранилище приложения
 */
export const store = configureStore({
  reducer: {
    error: errorReducer,
    user: userReducer,
    location: locationReducer,
  },
});

export const createStore = () => {
  const serverState = window.__PRELOADED_STATE__ || {};

  return configureStore({
    reducer: {
      error: errorReducer,
      user: userReducer,
      location: locationReducer,
    },
    preloadedState: serverState,
  });
};

/**
 * Тип стейта приложения
 */
export type RootState = ReturnType<typeof store.getState>;

/**
 * Тип диспетчера
 */
export type AppDispatch = typeof store.dispatch;
