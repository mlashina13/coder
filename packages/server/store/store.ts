// eslint-disable-next-line import/no-extraneous-dependencies
import {
  EnhancedStore,
  StoreEnhancer,
  ThunkDispatch,
  Tuple,
  UnknownAction,
  configureStore,
} from '@reduxjs/toolkit';
import { errorReducer, locationReducer } from './slices';

export type StoreType = EnhancedStore<
  Record<string, unknown>,
  UnknownAction,
  Tuple<
    [
      StoreEnhancer<{
        dispatch: ThunkDispatch<Record<string, unknown>, undefined, UnknownAction>;
      }>,
      StoreEnhancer
    ]
  >
>;

/**
 * Хранилище приложения
 */
export const store: StoreType = configureStore({
  reducer: {
    error: errorReducer,
    // user: userReducer,
    location: locationReducer,
  },
});

/**
 * Тип стейта приложения
 */
export type RootState = ReturnType<typeof store.getState>;

/**
 * Тип диспетчера
 */
export type AppDispatch = typeof store.dispatch;
