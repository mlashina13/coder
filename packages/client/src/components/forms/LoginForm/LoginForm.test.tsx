/* eslint-disable import/no-extraneous-dependencies */
/**
 * @jest-environment node
 */
import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { store } from '../../../store';
import { LoginForm } from './LoginForm';

configure({ adapter: new Adapter() });
describe('Login Form', () => {
  // TODO: временно убираем, вернем в следующих задачах
  // test('Должен верно отрисовываться', () => {
  //   const WrappedForm = () => (
  //     <Provider store={store}>
  //       <BrowserRouter>
  //         <LoginForm />
  //       </BrowserRouter>
  //     </Provider>
  //   );

  //   const tree = render(<WrappedForm />);
  //   expect(tree).toMatchSnapshot();
  // });

  describe('Проверка ф-ти', () => {
    test('Должен обновлять поле с паролем', () => {
      const WrappedForm = () => (
        <Provider store={store}>
          <BrowserRouter>
            <LoginForm />
          </BrowserRouter>
        </Provider>
      );

      const passwordInput = mount<React.FC>(<WrappedForm />)
        .find(LoginForm)
        .find('input[name="password"]');

      act(() => {
        passwordInput.simulate('change', {
          persist: () => undefined,
          target: {
            name: 'password',
            value: '12345678',
          },
        });
      });
      expect(passwordInput.html()).toMatch('12345678');
    });

    test('Должен обновлять поле с логином', () => {
      const WrappedForm = () => (
        <Provider store={store}>
          <BrowserRouter>
            <LoginForm />
          </BrowserRouter>
        </Provider>
      );

      const loginInput = mount<React.FC>(<WrappedForm />)
        .find(LoginForm)
        .find("input[name='login']");

      act(() => {
        loginInput.simulate('change', {
          persist: () => undefined,
          target: {
            name: 'login',
            value: 'newName',
          },
        });
      });

      expect(loginInput.html()).toMatch('newName');
    });
  });
});
