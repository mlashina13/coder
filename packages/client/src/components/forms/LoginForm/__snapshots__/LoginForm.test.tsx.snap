// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Должен верно отрисовываться 1`] = `
LoadedCheerio {
  "0": <form
    class="login-form"
  >
    <div
      class="MuiFormControl-root MuiTextField-root input login-form__input requiredField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":R1:"
        id=":R1:-label"
      >
        Логин
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1ff8729-MuiInputBase-root-MuiFilledInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-c5v1qu-MuiInputBase-input-MuiFilledInput-input"
          id=":R1:"
          name="login"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root input login-form__input requiredField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":R2:"
        id=":R2:-label"
      >
        Пароль
      </label>
      <div
        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-1ff8729-MuiInputBase-root-MuiFilledInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="off"
          class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-c5v1qu-MuiInputBase-input-MuiFilledInput-input"
          id=":R2:"
          name="password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium button css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Войти
    </button>
  </form>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
